src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -797,7 +815,68 @@ private void ImportCall(ILOpcode opcode, int token)
                 throw new NotImplementedException();
             }
 
-            HandleCall(callee);
+            if (callee.OwningType.IsDelegate)
+            {
+                throw new NotImplementedException();
+            }
+
+            HandleCall(callee, opcode);
+        }
+
+        private LLVMValueRef LLVMFunctionForMethod(MethodDesc callee, StackEntry thisPointer, bool isCallVirt)
+        {
+            string calleeName = _compilation.NameMangler.GetMangledMethodName(callee).ToString();
+            if (thisPointer != null && callee.IsVirtual && isCallVirt)
+            {
+                // TODO: Full resolution of virtual methods
+                if (!callee.IsNewSlot)
+                    throw new NotImplementedException();
+
+                if (!_compilation.HasFixedSlotVTable(callee.OwningType))
+                    _dependencies.Add(_compilation.NodeFactory.VirtualMethodUse(callee));
+
+                //TODO: needs runtime support for DispatchByInterface
+                if (callee.OwningType.IsInterface)
+                    throw new NotImplementedException();
+
+                return GetVirtualSlot(thisPointer.ValueAsType(LLVM.PointerType(LLVM.Int8Type(), 0), _builder), callee);
+            }
+            else
+            {
+                return GetOrCreateLLVMFunction(calleeName);
+            }
+        }
+
+        private LLVMValueRef GetOrCreateMethodSlot(MethodDesc method)
+        {
+            var globalRefName = "__getslot__" + _compilation.NameMangler.GetMangledMethodName(method);
+            LLVMValueRef slot = LLVM.GetNamedGlobal(Module, globalRefName);
+            if(slot.Pointer == IntPtr.Zero)
+            {
+                slot = LLVM.AddGlobal(Module, LLVM.Int32Type(), globalRefName);
+            }
+            return LLVM.BuildLoad(_builder, slot, string.Empty);
+        }
+
+        private LLVMValueRef GetVirtualSlot(LLVMValueRef objectPtr, MethodDesc method)
+        {
+            Debug.Assert(method.IsVirtual);
+
+            LLVMValueRef slot = GetOrCreateMethodSlot(method);
+            if (method.OwningType.IsInterface)
```

Comment:
My understanding of the interface dispatch was that the helper functions ended up just returning the target function pointer, So I was planning on emitting the runtime call inside that if and returning the result as though it had come from a normal virtual slot. Perhaps I should rename this method to something else.

Reply:
Renaming isn't a bad idea since slots are specific to virtuals. For interfaces, we'll have two choices:
1. Call RhpResolveInterfaceMethod, which returns the target function pointer
2. Set up stubs that look like https://github.com/dotnet/corert/blob/master/src/Native/Runtime/amd64/StubDispatch.S that tail call to the target. This is slightly better performance, but will take extra work to build.